<a href="https://colab.research.google.com/github/olodenm/info_2022_olo/blob/Netology_HW/%D0%94%D0%BE%D0%BC%D0%B0%D1%88%D0%BD%D0%B5%D0%B5_%D0%B7%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5_%D0%BA_%D0%BB%D0%B5%D0%BA%D1%86%D0%B8%D0%B8_%C2%AB%D0%91%D0%B0%D0%B7%D0%BE%D0%B2%D1%8B%D0%B5_%D0%BF%D0%BE%D0%BD%D1%8F%D1%82%D0%B8%D1%8F_%D1%81%D1%82%D0%B0%D1%82%D0%B8%D1%81%D1%82%D0%B8%D0%BA%D0%B8%C2%BB.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Задание 1. Загрузка данных
Изучить представленный [набор данных](/content/horse_data.csv) на основе описания его столбцов, загрузить его и оставить 8 столбцов для дальнейшего изучения: **surgery?, Age, rectal temperature, pulse, respiratory rate, temperature of extremities, pain, outcome.**

In [ ]:
import pandas as pd
import numpy as np
import random as r

In [ ]:
# скачиваем исходный df
! wget https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv

--2023-10-27 08:49:22--  https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.110.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 25151 (25K) [text/plain]
Saving to: ‘horse_data.csv.1’

horse_data.csv.1    100%[===================>]  24.56K  --.-KB/s    in 0.002s  

2023-10-27 08:49:22 (10.3 MB/s) - ‘horse_data.csv.1’ saved [25151/25151]



In [ ]:
# читаем файл csv без использования первой строки в качестве заголовка
# загружаем df используя индексы колонок, которые хотим читать
# задаем название колонок согласно документации и учитываем все пропуски в df
df = pd.read_csv('/content/horse_data.csv',
                  header=None,
                  usecols=[0, 1, 3, 4, 5, 6, 10, 22 ],
                  names=['surgery?', 'age', 'rectal_temperature', 'pulse',
                         'respiratory_rate', 'temperature_of_extremities',
                         'pain', 'outcome'],
                  na_values=['?'])
df.head()

surgery?  age  rectal_temperature  pulse  respiratory_rate  \
0       2.0    1                38.5   66.0              28.0   
1       1.0    1                39.2   88.0              20.0   
2       2.0    1                38.3   40.0              24.0   
3       1.0    9                39.1  164.0              84.0   
4       2.0    1                37.3  104.0              35.0   

   temperature_of_extremities  pain  outcome  
0                         3.0   5.0      2.0  
1                         NaN   3.0      3.0  
2                         1.0   3.0      1.0  
3                         4.0   2.0      2.0  
4                         NaN   NaN      2.0

#Задание 2. Первичное изучение данных
Проанализировать значения по столбцам, рассчитать базовые статистики, найти выбросы.



In [ ]:
# Заменим значение "9" на 2, согласно документации (.names)
df['age'] = df['age'].replace(9, 2)
# смотрим значенмя в стоблце
df['age'].value_counts()

1    276
2     24
Name: age, dtype: int64

In [ ]:
# смотрим общую информацию о df
# (не знаю, нужно ли переводить все столбцы к одному Dtype (?))
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    299 non-null    float64
 1   age                         300 non-null    int64  
 2   rectal_temperature          240 non-null    float64
 3   pulse                       276 non-null    float64
 4   respiratory_rate            242 non-null    float64
 5   temperature_of_extremities  244 non-null    float64
 6   pain                        245 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


In [ ]:
# базовые статистики для непрерывных величин
basic_statistics = df[['rectal_temperature', 'pulse','respiratory_rate']].describe()

# мода для категориальных величин
moda = df[['surgery?', 'age', 'temperature_of_extremities', 'pain', 'outcome']].mode()

# Анализ распределения по категориям:
values = df[['surgery?', 'age', 'temperature_of_extremities', 'pain', 'outcome']].value_counts()

#Подсчет количества пропусков в каждом столбце
missing_values = df.isnull().sum()

# Проверка уникальных значений на предмет ошибок
unique_surgery = df['surgery?'].unique()
unique_age = df['age'].unique()
unique_temperature = df['temperature_of_extremities'].unique()
unique_pain = df['pain'].unique()
unique_outcome = df['outcome'].unique()
unique_values = (unique_surgery, unique_age, unique_temperature, unique_pain, unique_outcome)

# медиана для каждого столбца
columns = ['rectal_temperature', 'pulse', 'respiratory_rate']
medians = df[columns].median()

# выбросы
q1 = df[columns].quantile(0.25)
q3 = df[columns].quantile(0.75)
iqr = q3 - q1
outliers = (df[columns] < (q1 - 1.5 * iqr)) | (df[columns] > (q3 + 1.5 * iqr))

result = f"Basic Statistics:\n{basic_statistics}\n\nMedians:\n{medians}\n\nModa:\n{moda}\n\nValues:\n{values}\n\nMissing_values:\n{missing_values}\n\nUnique colums values:\n{unique_values}\n\nOutliers:\n{outliers.sum()}"
print(result)

Basic Statistics:
       rectal_temperature       pulse  respiratory_rate
count          240.000000  276.000000        242.000000
mean            38.167917   71.913043         30.417355
std              0.732289   28.630557         17.642231
min             35.400000   30.000000          8.000000
25%             37.800000   48.000000         18.500000
50%             38.200000   64.000000         24.500000
75%             38.500000   88.000000         36.000000
max             40.800000  184.000000         96.000000

Medians:
rectal_temperature    38.2
pulse                 64.0
respiratory_rate      24.5
dtype: float64

Moda:
   surgery?  age  temperature_of_extremities  pain  outcome
0       1.0    1                         3.0   3.0      1.0

Values:
surgery?  age  temperature_of_extremities  pain  outcome
1.0       1    1.0                         3.0   1.0        13
2.0       1    1.0                         1.0   1.0        12
               2.0                         1.0   1.0 

In [ ]:
# Проведем анализ данных с учетом пропусков. Цель анализа:
#  чем закончилось лечение лошадей в (%) взрослых и молодых,
#  у которых была операция или операции не было
# разделим df на четыре новых df по столбцу 'age' и 'surgery?'
#  1) в первом старые лошади (1) прощедшие операцию (1),
#  2) во втором старые лошади(1) без операции (2),
#  3) в третьем молодые лошади (2) прощедшие операцию (1),
#  4) в четвертом молодые лошади (2) без операции (2)

total_old_horse_sur = df[(df['age'] == 1) & (df['surgery?'] == 1)]['outcome'].count()
total_old_horse_nosur = df[(df['age'] == 1) & (df['surgery?'] == 2)]['outcome'].count()
total_young_horse_sur = df[(df['age'] == 2) & (df['surgery?'] == 1)]['outcome'].count()
total_young_horse_nosur = df[(df['age'] == 2) & (df['surgery?'] == 2)]['outcome'].count()

# Подсчет процентного соотношения каждой категории внутри группы
df_old_horse_percent_sur = (df[(df['age'] == 1) & (df['surgery?'] == 1)].groupby('outcome')['surgery?'].count() / total_old_horse_sur) * 100
df_old_horse_percent_nosur = (df[(df['age'] == 1) & (df['surgery?'] == 2)].groupby('outcome')['surgery?'].count() / total_old_horse_nosur) * 100
df_young_horse_percent_sur = (df[(df['age'] == 2) & (df['surgery?'] == 1)].groupby('outcome')['surgery?'].count() / total_young_horse_sur) * 100
df_young_horse_percent_nosur = (df[(df['age'] == 2) & (df['surgery?'] == 2)].groupby('outcome')['surgery?'].count() / total_young_horse_nosur) * 100

# Разница между группами в процентах
difference_percent_sur = (df_old_horse_percent_sur - df_young_horse_percent_sur)
difference_percent_nosur = (df_old_horse_percent_nosur - df_young_horse_percent_nosur)

result = f"""
1)Старые лошади прощедшие операцию:
{df_old_horse_percent_sur}\n
Молодые лошади прощедщие операцию:
{df_young_horse_percent_sur}\n
Старые лошади без операции:
{df_old_horse_percent_nosur}\n
Молодые лошади без операции::
{df_young_horse_percent_nosur}\n
Разница в процентах между старыми и молодыми лодащьми прощедщими операцию:
{difference_percent_sur.reset_index()}\n
Разница в процентах между старыми и молодыми лодащьми без операции:
{difference_percent_nosur.reset_index()}\n
"""
print(result)


1)Старые лошади прощедшие операцию:
outcome
1.0    54.320988
2.0    29.629630
3.0    16.049383
Name: surgery?, dtype: float64

Молодые лошади прощедщие операцию:
outcome
1.0    38.888889
2.0    55.555556
3.0     5.555556
Name: surgery?, dtype: float64

Старые лошади без операции:
outcome
1.0    69.026549
2.0    15.929204
3.0    15.044248
Name: surgery?, dtype: float64

Молодые лошади без операции::
outcome
1.0    83.333333
2.0    16.666667
Name: surgery?, dtype: float64

Разница в процентах между старыми и молодыми лодащьми прощедщими операцию:
   outcome   surgery?
0      1.0  15.432099
1      2.0 -25.925926
2      3.0  10.493827

Разница в процентах между старыми и молодыми лодащьми без операции:
   outcome   surgery?
0      1.0 -14.306785
1      2.0  -0.737463
2      3.0        NaN




#Задание 3. Работа с пропусками
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе заполнения пропусков по каждому столбцу на основе рассчитанных статистик и возможной взаимосвязи значений в них. Сформировать датафрейм, в котором пропуски будут отсутствовать.

In [ ]:
df_nopass = df.copy()

In [ ]:
# Расчет количества пропусков в каждом столбце
df_nopass.isna().sum()

surgery?                      0
age                           0
rectal_temperature            0
pulse                         0
respiratory_rate              0
temperature_of_extremities    0
pain                          0
outcome                       0
dtype: int64

In [ ]:
#  Заполнение пропусков для категориальных столбцов

# 1)столбец "surgery" пропуск заполнил 1. самое частое значение (276)
df_nopass['surgery?'].fillna((1), inplace=True)

# 2). temperature of extremities заполняю модой.
df_nopass['temperature_of_extremities'].fillna(df_nopass['temperature_of_extremities'].mode()[0], inplace=True)

# 3) pain заполняю модой.
df_nopass['pain'].fillna(df_nopass['pain'].mode()[0], inplace=True)

# 4) Заполним столбец самым частым значением (lived)
df_nopass['outcome'].fillna((1), inplace=True)

# ? А если мы хотим заполнить  заполнить пропуски в столбце например
#'temperature_of_extremities', чтобы понять как лучше заполнить пропуски
# можно расчитать кореляцию с другим столбцом 'rectal_temperature',
# Например, при повышенной ректальной температуре у лошади
# 'temperature_of_extremities' обычно "2.0" (Warm)
#но при этом чтобы узнать карреляцию не должно быть пропусков?

In [ ]:
# Заполнение пропусков в количественных столбцов

# 1) Заполнил пропуски согласно документации нормальным значением (normal temp is 37.8)
df_nopass['rectal_temperature'].fillna((37.8), inplace=True)

# 2)Согласно документации нормальный пульс для взрослой лошади (surgery == 1)
# между 30-40 упростим задачу и запишем среднее значение (37.85)
average_pulse_surgery_1 = round(df_nopass[(df_nopass['pulse'] >= 30) &
                                     (df_nopass['pulse'] <= 40) &
                                     (df_nopass['surgery?'] == 1)]['pulse'].mean(), 2)
df_nopass.loc[(df_nopass['pulse'].isna()) &
               (df_nopass['surgery?'] == 1), 'pulse'] = average_pulse_surgery_1
# оставшиеся пропуски заполнил медианой
df_nopass['pulse'].fillna(df_nopass['pulse'].median(), inplace=True)

# 3) respiratory_rate согласно документации нормальное дыхание(normal rate is 8 to 10)
# Генерируем случайные значения от 8 до 10
random_values = [r.randint(8, 10) if pd.isna(val) else val for val in df_nopass['respiratory_rate']]
# заполнил столбец "respiratory_rate" новыми значениями
df_nopass['respiratory_rate'] = random_values

In [ ]:
# Выбросы (outlers)
# имеется 5 выбросов в столбце pulse
df_nopass[df_nopass['pulse'] > 149]

# выбросы в столбце pulse идут по верхнему значению, если посмотреть df
# то мы можем соотнести их к тому что операция прошла не успешно,
# 80% лошадей с повышеным пульсом и прощедщими операцию скончались
# так же у них была и повышеная темпиратура

surgery?  age  rectal_temperature  pulse  respiratory_rate  \
3         1.0    2                39.1  164.0              84.0   
41        2.0    2                39.0  150.0              72.0   
55        1.0    2                38.6  160.0              20.0   
255       1.0    2                38.8  184.0              84.0   
275       1.0    2                38.8  150.0              50.0   

     temperature_of_extremities  pain  outcome  
3                           4.0   2.0      2.0  
41                          3.0   3.0      1.0  
55                          3.0   3.0      2.0  
255                         1.0   4.0      2.0  
275                         1.0   5.0      2.0

In [ ]:
# Частота дыхания
df_nopass[df_nopass['respiratory_rate'] > 75]

# по df видно что у всех лошадей с значениями выброса в столбце
# были операции, можем предполодить что лошади боялись операции,
# Возможно лучше было подготовить лощадь к операции и успокоить
# Например дополнительным успокоительным и тд.

surgery?  age  rectal_temperature  pulse  respiratory_rate  \
3         1.0    2                39.1  164.0              84.0   
39        1.0    2                39.2  146.0              96.0   
82        1.0    2                38.1  100.0              80.0   
84        1.0    1                37.8   60.0              80.0   
106       1.0    1                38.3   52.0              96.0   
186       1.0    1                39.3   64.0              90.0   
208       1.0    1                37.8   88.0              80.0   
244       1.0    2                38.2  124.0              88.0   
255       1.0    2                38.8  184.0              84.0   
269       1.0    1                39.3   64.0              90.0   

     temperature_of_extremities  pain  outcome  
3                           4.0   2.0      2.0  
39                          3.0   3.0      2.0  
82                          3.0   3.0      1.0  
84                          1.0   2.0      1.0  
106                         3.0   3.0      1.0  
186                         2.0   3.0      1.0  
208                         3.0   3.0      3.0  
244                         1.0   2.0      1.0  
255                         1.0   4.0      2.0  
269                         2.0   3.0      1.0

In [ ]:
correlation_matrix = df_nopass[['rectal_temperature', 'pulse', 'respiratory_rate']].corr()
correlation_matrix
# Корреляция между "pulse" и "respiratory_rate" составляет примерно 0.470,
# что указывает на более сильную положительную корреляцию между
# этими двумя столбцами, в сравнение с другими столбцами
# Это может означать, что у лошадей с более высоким пульсом  частота дыхания чаще.

rectal_temperature     pulse  respiratory_rate
rectal_temperature            1.000000  0.168213          0.222628
pulse                         0.168213  1.000000          0.401771
respiratory_rate              0.222628  0.401771          1.000000

In [ ]:
# DF без пропусков
df_nopass

surgery?  age  rectal_temperature  pulse  respiratory_rate  \
0         2.0    1                38.5   66.0              28.0   
1         1.0    1                39.2   88.0              20.0   
2         2.0    1                38.3   40.0              24.0   
3         1.0    2                39.1  164.0              84.0   
4         2.0    1                37.3  104.0              35.0   
..        ...  ...                 ...    ...               ...   
295       1.0    1                37.8  120.0              70.0   
296       2.0    1                37.2   72.0              24.0   
297       1.0    1                37.5   72.0              30.0   
298       1.0    1                36.5  100.0              24.0   
299       1.0    1                37.2   40.0              20.0   

     temperature_of_extremities  pain  outcome  
0                           3.0   5.0      2.0  
1                           3.0   3.0      3.0  
2                           1.0   3.0      1.0  
3                           4.0   2.0      2.0  
4                           3.0   3.0      2.0  
..                          ...   ...      ...  
295                         4.0   2.0      3.0  
296                         3.0   4.0      3.0  
297                         4.0   4.0      2.0  
298                         3.0   3.0      1.0  
299                         3.0   3.0      3.0  

[300 rows x 8 columns]

In [ ]:
# Проведем анализ на данных без учета пропусков.Цель анализа:
#  чем закончилось лечение лошадей в (%) взрослых и молодых,
#   у которых была операция или операции не было
# разделим df на четыре новых df по столбцу 'age' и 'surgery?'
#  1) в первом старые лошади (1) прощедшие операцию (1),
#  2) во втором старые лошади(1) без операции (2),
#  3) в третьем молодые лошади (2) прощедшие операцию (1),
#  4) в четвертом молодые лошади (2) без операции (2)

total_old_horse_sur_nopass = df_nopass[(df_nopass['age'] == 1) & (df_nopass['surgery?'] == 1)]['outcome'].count()
total_old_horse_nosur_nopass = df_nopass[(df_nopass['age'] == 1) & (df_nopass['surgery?'] == 2)]['outcome'].count()
total_young_horse_sur_nopass = df_nopass[(df_nopass['age'] == 2) & (df_nopass['surgery?'] == 1)]['outcome'].count()
total_young_horse_nosur_nopass = df_nopass[(df_nopass['age'] == 2) & (df_nopass['surgery?'] == 2)]['outcome'].count()

# Подсчет процентного соотношения каждой категории внутри группы
df_old_horse_percent_sur_nopass = (df_nopass[(df_nopass['age'] == 1) & (df_nopass['surgery?'] == 1)].groupby('outcome')['surgery?'].count() / total_old_horse_sur_nopass) * 100
df_old_horse_percent_nosur_nopass = (df_nopass[(df_nopass['age'] == 1) & (df_nopass['surgery?'] == 2)].groupby('outcome')['surgery?'].count() / total_old_horse_nosur_nopass) * 100
df_young_horse_percent_sur_nopass = (df_nopass[(df_nopass['age'] == 2) & (df_nopass['surgery?'] == 1)].groupby('outcome')['surgery?'].count() / total_young_horse_sur_nopass) * 100
df_young_horse_percent_nosur_nopass = (df_nopass[(df_nopass['age'] == 2) & (df_nopass['surgery?'] == 2)].groupby('outcome')['surgery?'].count() / total_young_horse_nosur_nopass) * 100

# Разница между группами в процентах
difference_percent_sur_nopass = (df_old_horse_percent_sur_nopass - df_young_horse_percent_sur_nopass)
difference_percent_nosur_nopass = (df_old_horse_percent_nosur_nopass - df_young_horse_percent_nosur_nopass)

result_nopass = f"""
1)Старые лошади прощедшие операцию:
{df_old_horse_percent_sur_nopass}\n
Молодые лошади прощедщие операцию:
{df_young_horse_percent_sur_nopass}\n
Старые лошади без операции:
{df_old_horse_percent_nosur_nopass}\n
Молодые лошади без операции::
{df_young_horse_percent_nosur_nopass}\n
Разница в процентах между старыми и молодыми лодащьми прощедщими операцию:
{difference_percent_sur_nopass.reset_index()}\n
Разница в процентах между старыми и молодыми лодащьми без операции:
{difference_percent_nosur_nopass.reset_index()}\n
"""
print(result_nopass)


1)Старые лошади прощедшие операцию:
outcome
1.0    54.601227
2.0    29.447853
3.0    15.950920
Name: surgery?, dtype: float64

Молодые лошади прощедщие операцию:
outcome
1.0    38.888889
2.0    55.555556
3.0     5.555556
Name: surgery?, dtype: float64

Старые лошади без операции:
outcome
1.0    69.026549
2.0    15.929204
3.0    15.044248
Name: surgery?, dtype: float64

Молодые лошади без операции::
outcome
1.0    83.333333
2.0    16.666667
Name: surgery?, dtype: float64

Разница в процентах между старыми и молодыми лодащьми прощедщими операцию:
   outcome   surgery?
0      1.0  15.712338
1      2.0 -26.107703
2      3.0  10.395365

Разница в процентах между старыми и молодыми лодащьми без операции:
   outcome   surgery?
0      1.0 -14.306785
1      2.0  -0.737463
2      3.0        NaN




 Для столбца "surgery?" и исхода 1 (выживание):
 Разница для старых лошадей прошедших операцию: 54.320988 - 54.601227 ≈ -0.280239
 примерно 0.28%

 Для столбца "surgery?" и исхода 2 (смерть):
 Разница для старых лошадей прошедших операцию: 29.629630 - 29.447853 ≈ 0.181777
 примерно 0.18%

 Для столбца "surgery?" и исхода 3 (усыпление):
 Разница для старых лошадей прошедших операцию: 16.049383 - 15.950920 ≈ 0.098463
 примерно 0.10%

 Общий вывод:
 Таким образом, мы видим не существенную разницу в значениях после заполнения пропусков
 только для старых лошадей прошедших операцию
 может сказать что больше значений было именно этих данных
